### PR TITLE
[ML] Stream Cohere Completion

### PR DESCRIPTION
**Pull request by prwhelan**
Implement and enable streaming for Cohere chat completions (v1).

Includes processor for ND JSON streaming responses.

### CODE REVIEW

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/cohere/CohereStreamingProcessor.java [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+        for (String json : item) {
+            try (var jsonParser = jsonParser(json)) {
+                var responseMap = jsonParser.map();
+                var eventType = (String) responseMap.get("event_type");
+                switch (eventType) {
+                    case "text-generation" -> parseText(responseMap).ifPresent(results::offer);
+                    case "stream-end" -> validateResponse(responseMap);
+                    case "stream-start", "search-queries-generation", "search-results", "citation-generation", "tool-calls-generation",
+                        "tool-calls-chunk" -> {
+                        // skip
```
Comment (maxhniebergall) [inline x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/cohere/CohereStreamingProcessor.java:48]: Should we log if this happens or anything?

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/response/streaming/NewlineDelimitedByteProcessor.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+/**
+ * Processes HttpResult bytes into lines separated by newlines, delimited by either line-feed or carriage-return line-feed.
+ * Downstream is responsible for validating the structure of the lines after they have been separated.
+ * Because Upstream (Apache) can send us a single line split between two HttpResults, this processor will aggregate bytes from the last
+ * HttpResult and append them to the front of the next HttpResult.
+ * When onComplete is called, the last batch is always flushed to the downstream onNext.
+ */
+public class NewlineDelimitedByteProcessor extends DelegatingProcessor<HttpResult, Deque<String>> {
+    private static final Pattern END_OF_LINE_REGEX = Pattern.compile("\\n|\\r\\n");
```
Comment (maxhniebergall) [inline x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/response/streaming/NewlineDelimitedByteProcessor.java:26]: thanks for pulling this out as a named constant!

**Review by maxhniebergall — APPROVED** [review]
LGTM

**Comment by prwhelan** [comment]
@elasticmachine update branch

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/cohere/CohereStreamingProcessor.java (hunk 1 above)
Comment (prwhelan) [reply x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/cohere/CohereStreamingProcessor.java:48]: We should add a debug log - these are included in most responses, and I'm not sure what to do with them or if they're even needed in our response object (for now

**Comment by prwhelan** [comment]
@elasticmachine update branch

**Pull request merged**